### PR TITLE
Fixed remaining bytebuffer wrapper

### PR DESCRIPTION
**Pull request by g4s8**
#43 - mark and reset `ByteBuffer` before and after reading from it

### CODE REVIEW

**Comment by paulodamaso** [comment]
@g4s8 I can review this one, please assign me

**Comment by g4s8** [comment]
@0crat assign @paulodamaso

**Comment by 0crat** [comment]
> @0crat assign @paulodamaso  ([here](https://github.com/artipie/asto/issues/44#issuecomment-584119410))

@g4s8 This pull request [#44](https://github.com/artipie/asto/issues/44) is assigned to @paulodamaso[/z](https://www.0crat.com/u/paulodamaso), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/fd5a6dd9-0ea8-4120-b264-4dc117d218a2); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

**Comment by 0crat** [comment]
Job [#44](https://github.com/artipie/asto/issues/44) is already in scope

**Comment by 0crat** [comment]
Manual assignment of issues is discouraged, see [§19](http://www.zerocracy.com/policy.html#19): -5 point(s) just awarded to @g4s8[/z](https://www.0crat.com/u/g4s8)

File: src/test/java/com/artipie/asto/RemainingTest.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+public final class RemainingTest {
+
+    @Test
+    public void readTwice() throws Exception {
+        final ByteBuffer buf = ByteBuffer.allocate(32);
+        final byte[] array = new byte[]{1, 2, 3, 4};
+        buf.put(array);
+        buf.flip();
+        MatcherAssert.assertThat(
+            "unexpected remaining array", new Remaining(buf).bytes(), Matchers.equalTo(array)
```
Comment (paulodamaso) [inline src/test/java/com/artipie/asto/RemainingTest.java]: @g4s8 Let's use `new IsEqual` instead static matchers. Also, this assertion failure message is unclear and should start with a capital letter

File: src/test/java/com/artipie/asto/RemainingTest.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        );
+        MatcherAssert.assertThat(
+            "failed to read array twice", new Remaining(buf).bytes(), Matchers.equalTo(array)
```
Comment (paulodamaso) [inline src/test/java/com/artipie/asto/RemainingTest.java]: @g4s8  Let's use `IsEqual`  instead static matchers. Also, this assertion failure message should start with a capital letter

File: src/test/java/com/artipie/asto/RemainingTest.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+ * @since 1.0
+ */
+public final class RemainingTest {
+
+    @Test
+    public void readTwice() throws Exception {
+        final ByteBuffer buf = ByteBuffer.allocate(32);
+        final byte[] array = new byte[]{1, 2, 3, 4};
+        buf.put(array);
+        buf.flip();
```
Comment (paulodamaso) [inline src/test/java/com/artipie/asto/RemainingTest.java]: @g4s8 We shouldn't have two assertions in the same test method, seems that we are testing two different things here

Comment (g4s8) [reply src/test/java/com/artipie/asto/RemainingTest.java]: @paulodamaso how would you test the bug #43 then? I created this test to reproduce that this class returns different byte array on second read.

**Comment by g4s8** [comment]
@paulodamaso fixed and answered, please check new commit.

> Shouldn't it at least implement Itrerable<Byte> or something like that, since it is a collection of bytes

I don't think I need to change this class in scope of bug fixing from #43

**Comment by 0crat** [comment]
Job was finished in 17 hours, bonus for fast delivery is possible (see [§36](http://www.zerocracy.com/policy.html#36))

**Comment by 0crat** [comment]
Order was finished: +20 point(s) just awarded to @paulodamaso[/z](https://www.0crat.com/u/paulodamaso)

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @g4s8[/z](https://www.0crat.com/u/g4s8)

**Pull request merged**